src/backend/x11/connection.rs

```diff
@@ -0,0 +1,135 @@
+    fn from(inner: ConnectError) -> Self {
+        Self::Xcb(inner)
+    }
+}
+
+/// A connection to the X server.
+///
+/// This contains a way to access both the xcb connection and the xlib Display.
+pub struct XConnection {
+    xlib_library: Xlib,
+    /// The xlib Display that initiated the XCBConnection
+    ///
+    /// If we want to allow creation of an EGL Context using X11, the XCB extensions are not
+    /// implemented most drivers yet. So we also expose the Xlib types so an OpenGL context
+    /// can be made in nearly every driver.
+    xlib_display: *mut x11_dl::xlib::Display,
+    xcb_connection: XCBConnection,
+}
+
+impl XConnection {
+    /// Attempts to connect to the X server.
+    pub(crate) fn new(_logger: &Logger) -> Result<(XConnection, usize), ConnectToXError> {
+        // TODO: Log setup process
+
+        let xlib = Xlib::open().expect("Failed to open xlib library");
+        let xlib_xcb = Xlib_xcb::open().expect("Failed to load xlib_libxcb");
+        let (display, screen_number) = unsafe {
+            let display = (xlib.XOpenDisplay)(ptr::null());
+
+            if display.is_null() {
+                return Err(ConnectToXError::XlibNoConnection);
+            }
+
+            (display, (xlib.XDefaultScreen)(display))
+        };
+
+        // Transfer ownership of the event queue to XCB since we use XCB to handle events.
+        let xcb_connection_t = unsafe {
+            let ptr = (xlib_xcb.XGetXCBConnection)(display);
+            (xlib_xcb.XSetEventQueueOwner)(display, XEventQueueOwner::XCBOwnsEventQueue);
+            ptr
+        };
+
+        let xcb_connection = unsafe {
+            // Xlib implementation does not use XCB
+            if xcb_connection_t.is_null() {
+                (xlib.XCloseDisplay)(display);
+                return Err(ConnectToXError::NoXlibXcb);
+            }
+
+            // Do not drop the connection upon closure since Xlib created the xcb_connection_t.
+            // The Drop impl of `XConnection` will shutdown the Xlib Display
+            XCBConnection::from_raw_xcb_connection(xcb_connection_t, false)
+        };
+
+        if xcb_connection.is_err() {
+            unsafe { (xlib.XCloseDisplay)(display) };
+            return Err(ConnectToXError::Xcb(xcb_connection.err().unwrap()));
+        }
+
+        Ok((
+            XConnection {
+                xlib_library: xlib,
+                xlib_display: display,
+                xcb_connection: xcb_connection.unwrap(),
+            },
+            screen_number as usize,
+        ))
+    }
+
+    /// Returns the xcb connection
+    pub fn xcb_connection(&self) -> &XCBConnection {
+        &self.xcb_connection
+    }
+
+    /// Returns a pointer to the xlib display.
+    pub fn xlib_display(&self) -> *mut x11_dl::xlib::Display {
+        self.xlib_display
+    }
+}
+
+// Xlib and libxcb are both thread safe.
+unsafe impl Send for XConnection {}
+unsafe impl Sync for XConnection {}
```

Comment:
As @DemiMarie already wrote: No, you need to call `xlib::XInitThreads()` once first. I'd recommend a global variable containing something like a once_cell for this so that this call even only happens once when another `XConnection` is constructed later on.